packages/auth/src/app.ts

```diff
@@ -36,6 +37,67 @@ export interface AppContextData extends DefaultContext {
 
 export type AppContext = Koa.ParameterizedContext<DefaultState, AppContextData>
 
+interface ClientKeyContext extends AppContext {
+  clientKeyId: string
+}
+
+type GrantRequest<BodyT = never, QueryT = ParsedUrlQuery> = Omit<
+  ClientKeyContext['request'],
+  'body'
+> & {
+  body: BodyT
+  query: ParsedUrlQuery & QueryT
+}
+
+type GrantContext<BodyT = never, QueryT = ParsedUrlQuery> = Omit<
+  ClientKeyContext,
+  'request'
+> & {
+  request: GrantRequest<BodyT, QueryT>
+}
+
+type InteractionRequest<
+  BodyT = never,
+  QueryT = ParsedUrlQuery,
+  ParamsT = { [key: string]: string }
+> = Omit<AppContext['request'], 'body'> & {
+  body: BodyT
+  query: ParsedUrlQuery & QueryT
+  params: ParamsT
+}
+
+type InteractionContext<QueryT, ParamsT> = Omit<AppContext, 'request'> & {
+  request: InteractionRequest<QueryT, ParamsT>
+}
+
+type TokenRequest<BodyT = never> = Omit<AppContext['request'], 'body'> & {
+  body?: BodyT
+}
+
+type TokenContext<BodyT = never> = Omit<AppContext, 'request'> & {
+  request: TokenRequest<BodyT>
+}
+
+type ManagementRequest = Omit<AppContext['request'], 'params'> & {
+  params?: Record<'id', string>
+}
+
+type ManagementContext = Omit<AppContext, 'request'> & {
+  request: ManagementRequest
+}
+
+export type CreateContext<BodyT> = GrantContext<BodyT>
+export type ContinueContext<BodyT, QueryT> = GrantContext<BodyT, QueryT>
+
+export type StartContext<QueryT, ParamsT> = InteractionContext<QueryT, ParamsT>
+export type GetContext<ParamsT> = InteractionContext<never, ParamsT>
+export type ChooseContext<ParamsT> = InteractionContext<never, ParamsT>
+export type FinishContext<ParamsT> = InteractionContext<never, ParamsT>
+
+export type IntrospectContext<BodyT> = TokenContext<BodyT>
+export type RevokeContext = ManagementContext
+export type RotateContext = ManagementContext
+
```

Comment:
The calls to `createValidatorMiddleware` below should specify the type guarded context.
For example:
https://github.com/interledger/rafiki/blob/e067823ff1e0a1f9cb13b4b14587c52547db70b1/packages/auth/src/app.ts#L196-L206
should change to have
```
      createValidatorMiddleware<CreateContext>(openApi.authServerSpec, {
        path: '/',
        method: HttpMethod.POST
      })
```

That gets tricky without having the particular `BodyT` defined in this file.
One solution is to use:
https://github.com/interledger/rafiki/blob/e067823ff1e0a1f9cb13b4b14587c52547db70b1/packages/backend/src/app.ts#L112-L117
then have
```
      createValidatorMiddleware<Context<grantRoutes.create>>(openApi.authServerSpec, {
        path: '/',
        method: HttpMethod.POST
      })
```
in which case a lot of these contexts can be defined in their corresponding `routes.ts` file and be defined with the specific generic type (`BodyT`, etc)